pkg/controller/container-runtime-config/container_runtime_config_controller.go

```diff
@@ -284,6 +308,62 @@ func (ctrl *Controller) itmsConfDeleted(_ interface{}) {
 	ctrl.imgQueue.Add("openshift-config")
 }
 
+func (ctrl *Controller) clusterImagePolicyAdded(_ interface{}) {
+	if !ctrl.sigstoreAPIEnabled() {
+		return
+	}
+	ctrl.imgQueue.Add("openshift-config")
+}
+
+func (ctrl *Controller) clusterImagePolicyUpdated(_, _ interface{}) {
+	if !ctrl.sigstoreAPIEnabled() {
+		return
+	}
+	ctrl.imgQueue.Add("openshift-config")
+}
+
+func (ctrl *Controller) clusterImagePolicyDeleted(_ interface{}) {
+	if !ctrl.sigstoreAPIEnabled() {
+		return
+	}
+	ctrl.imagePoliciesDeleted()
```

Comment:
This was moved inside the sync loop, but it still seems coupled to the `clusterImagePolicyDeleted` event via the event key.

This still seems unexpected to me — the deletion code would not trigger on the periodic (20-minute?) resent, and it would not trigger on the sync when the controller first starts (there might have been no running controller to observe the “deleted” event).

That could all well be just my ignorance of a well-known pattern; if so, I’d appreciate a pointer.

Reply:
The current logic for generating mchineconfig ignition ignores the empty data. So if the `ClusterImagePolicy` got deleted, it's no-op to the policy machine config.  Even if the ignition can be updated/created, I think it will be an machine config without any data. https://github.com/openshift/machine-config-operator/blob/f799d7f337fbd7c00f24e4422192387bd1ca686f/pkg/controller/container-runtime-config/helpers.go#L135-L139
The matching config should be rendered by multiple objects, so we can not use the garbage collection feature via the OwnerReference field. So mange the machine config deletion by `clusterImagePolicyDeleted`. `ContainerRuntimrConfig` has something same as this https://github.com/openshift/machine-config-operator/blob/78a79676b21920b5d017060b437638a246602503/pkg/controller/container-runtime-config/container_runtime_config_controller.go#L317

Reply:
Considering the potential complication that the `imagePoliciesDeleted` can fail, we can just merge all the CIP configuration to current machine config 99-master/worker-generated-registries

Reply:
Thanks, that makes sense to me.